File: internal/service/overdue_svc.go
package service

import (
	"context"
	"log/slog"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
)

// overdueInvoiceRepo defines the invoice repository methods needed by OverdueService.
type overdueInvoiceRepo interface {
	ListOverdueCandidateIDs(ctx context.Context, beforeDate time.Time) ([]int64, error)
	UpdateStatus(ctx context.Context, id int64, status string) error
}

// statusHistoryRepo defines the status history repository methods needed by OverdueService.
type statusHistoryRepo interface {
	Create(ctx context.Context, change *domain.InvoiceStatusChange) error
	ListByInvoiceID(ctx context.Context, invoiceID int64) ([]domain.InvoiceStatusChange, error)
}

// OverdueService handles overdue detection and invoice status history.
type OverdueService struct {
	invoiceRepo overdueInvoiceRepo
	historyRepo statusHistoryRepo
}

// NewOverdueService creates a new OverdueService.
func NewOverdueService(invoiceRepo overdueInvoiceRepo, historyRepo statusHistoryRepo) *OverdueService {
	return &OverdueService{
		invoiceRepo: invoiceRepo,
		historyRepo: historyRepo,
	}
}

// CheckOverdue finds all sent invoices past their due date, marks them as overdue,
// and records the status change in history. Returns the number of invoices marked.
func (s *OverdueService) CheckOverdue(ctx context.Context) (int, error) {
	now := time.Now()

	// First, find all candidates before updating.
	ids, err := s.invoiceRepo.ListOverdueCandidateIDs(ctx, now)
	if err != nil {
		return 0, err
	}

	if len(ids) == 0 {
		return 0, nil
	}

	count := 0
	for _, id := range ids {
		if err := s.invoiceRepo.UpdateStatus(ctx, id, domain.InvoiceStatusOverdue); err != nil {
			slog.Error("failed to mark invoice as overdue", "invoice_id", id, "error", err)
			continue
		}
		count++

		change := &domain.InvoiceStatusChange{
			InvoiceID: id,
			OldStatus: domain.InvoiceStatusSent,
			NewStatus: domain.InvoiceStatusOverdue,
			ChangedAt: now,
			Note:      "automatically marked as overdue",
		}
		if err := s.historyRepo.Create(ctx, change); err != nil {
			slog.Error("failed to record status change for overdue invoice", "invoice_id", id, "error", err)
		}
	}

	return count, nil
}

// GetHistory returns the status change history for a given invoice.
func (s *OverdueService) GetHistory(ctx context.Context, invoiceID int64) ([]domain.InvoiceStatusChange, error) {
	return s.historyRepo.ListByInvoiceID(ctx, invoiceID)
}

File: rust/crates/zfaktury-core/src/service/overdue_svc.rs
use std::sync::Arc;

use zfaktury_domain::{DomainError, InvoiceStatusChange};

use crate::repository::traits::{InvoiceRepo, StatusHistoryRepo};

/// Service for overdue detection and invoice status history.
pub struct OverdueService {
    invoice_repo: Arc<dyn InvoiceRepo + Send + Sync>,
    history_repo: Arc<dyn StatusHistoryRepo + Send + Sync>,
}

impl OverdueService {
    pub fn new(
        invoice_repo: Arc<dyn InvoiceRepo + Send + Sync>,
        history_repo: Arc<dyn StatusHistoryRepo + Send + Sync>,
    ) -> Self {
        Self {
            invoice_repo,
            history_repo,
        }
    }

    /// Finds all sent invoices past due date and marks them as overdue.
    pub fn check_overdue(&self) -> Result<i32, DomainError> {
        use zfaktury_domain::{InvoiceFilter, InvoiceStatus};
        let now = chrono::Local::now().naive_local();

        // List all sent invoices and check due dates.
        let (invoices, _) = self.invoice_repo.list(&InvoiceFilter {
            status: Some(InvoiceStatus::Sent),
            limit: 10000,
            ..Default::default()
        })?;

        let today = chrono::Local::now().date_naive();
        let mut count = 0;

        for inv in &invoices {
            if inv.due_date < today {
                if let Err(e) = self.invoice_repo.update_status(inv.id, "overdue") {
                    log::error!("failed to mark invoice {} as overdue: {}", inv.id, e);
                    continue;
                }
                count += 1;

                let mut change = InvoiceStatusChange {
                    id: 0,
                    invoice_id: inv.id,
                    old_status: InvoiceStatus::Sent,
                    new_status: InvoiceStatus::Overdue,
                    changed_at: now,
                    note: "automatically marked as overdue".to_string(),
                };
                if let Err(e) = self.history_repo.create(&mut change) {
                    log::error!(
                        "failed to record status change for invoice {}: {}",
                        inv.id,
                        e
                    );
                }
            }
        }

        Ok(count)
    }

    /// Returns the status change history for a given invoice.
    pub fn get_history(&self, invoice_id: i64) -> Result<Vec<InvoiceStatusChange>, DomainError> {
        self.history_repo.list_by_invoice_id(invoice_id)
    }
}
